### DISCUSSION TITLE
Is there a way to pull back the binding names from the results in submitted order?

### DISCUSSION THREAD

**Discussion by nguyenm100** [eclipse-rdf4j/rdf4j#3583] [Q&A] [answered]
Say I have a query:
```
select ?s ?p ?o where { ?s ?p ?o }
```
and I execute the parsequery and evaluate functions.  upon receiving back the results, can I pull back the binded names (?s, ?p, ?o) in the order they were submitted?  Seems I get back a set of bindingnames which doesn't guarantee any iteration order.

If I'm looking to display the results in a table, it's a bit confusing if the columns are out of order from the initial query.

**Comment by abrokenjester** [comment, marked as answer]
[TupleQueryResult#getBindingNames](https://rdf4j.org/javadoc/latest/index.html?org/eclipse/rdf4j/query/TupleQueryResult.html) gives back the binding names, in projection order.

**Reply by nguyenm100** [reply]
Hi, follow up on this.  Didn't realize there were multiple ways to execute a SPARQL statement and it looks like to get TupleQueryResult, you need to go through the Repository API?  Per snippet below, I'm executing through the SAIL connection which returns back BindingSet which leads to my original dilemma (note: the documentation for SailConnection.evaluate() says the return should be a TupleQueryResult -- https://rdf4j.org/javadoc/3.4.3/org/eclipse/rdf4j/sail/SailConnection.html#evaluate-org.eclipse.rdf4j.query.algebra.TupleExpr-org.eclipse.rdf4j.query.Dataset-org.eclipse.rdf4j.query.BindingSet-boolean-).  Is it possible to get at the TupleQueryResult from this approach?

```
        try (SailConnection sc = sail.getConnection()) {

            SPARQLParser parser = new SPARQLParser();
            BindingSet bindings = new EmptyBindingSet();

            ParsedQuery query;
            CloseableIteration<? extends BindingSet, QueryEvaluationException> results;
            
            query = parser.parseQuery(queryStr, this.baseUri);
            results = sc.evaluate(query.getTupleExpr(), query.getDataset(), bindings, false);
            
            <snip>
```

**Reply by abrokenjester** [reply]
First of all: the SAIL API is really not designed for direct usage like this: it's a system-internal API of RDF4J that's only public to facilitate store vendors an implementation target. It's not suitable for direct usage in client code: that's what the Repository API is designed for. I'd strongly urge you to change your approach.

Having said all that: the javadoc for this SAIL method is misleading, SailConnection never gives back a TupleQueryResult. Thanks for pointing that out, I'll see about correcting it. 

You could probably construct a TupleQueryResult yourself from the parsed query and the result iterator, but to be honest you'd just be duplicating what the Repository API already handles for you. One way to get the binding names is to actually extract them from the parsed query itself: `query.getTupleExpr().getBindingNames()` ought to do the trick.

Again though: consider switching to using the Repository API, instead of parsing the query manually like this and executing directly against the SAIL. You'll save yourself a lot of headaches in the longer term.

**Reply by nguyenm100** [reply]
Ok thanks so much. I mentioned I was trying to get SAIL working atop Tinkerpop so working at the level. 

query.getTupleExpr().getBindingNames() still returns a set :-(

**Reply by abrokenjester** [reply]
Ah! I wasn't aware that's what you were doing. If you haven't seen it yet, perhaps https://github.com/tinkerpop/blueprints/wiki/Sail-Implementation might be of interest. It's an old Tinkerpop sail implementation. Out of date ( still based on Sesame) but it might be useful for inspiration?

**Reply by nguyenm100** [reply]
Hi, yes saw that.  This one seems more up to date https://github.com/joshsh/graphsail.  Trying to work through some performance issues with the way JG/TP does things.
